scipy/stats/_qmc.py

```diff
@@ -0,0 +1,1493 @@
+                             "n to be a power of 2. {0} points have been "
+                             "previously generated, then: n={0}+2**{1}={2}. "
+                             "If you still want to do this, the function "
+                             "'Sobol.random()' can be used."
+                             .format(self.num_generated, m, total_n_samples))
+
+        return self.random(n_samples=n_samples)
+
+    def reset(self):
+        """Reset the engine to base state.
+
+        Returns
+        -------
+        engine: Sobol
+            Engine reset to its base state.
+
+        """
+        self._quasi = self._shift.copy()
+        self.num_generated = 0
+        return self
+
+    def fast_forward(self, n):
+        """Fast-forward the sequence by `n` positions.
+
+        Parameters
+        ----------
+        n: int
+            Number of points to skip in the sequence.
+
+        Returns
+        -------
+        engine: Sobol
+            The fast-forwarded engine.
+
+        """
+        if self.num_generated == 0:
+            _fast_forward(n - 1, self.num_generated, self.dim,
+                          self._sv, self._quasi)
+        else:
+            _fast_forward(n, self.num_generated - 1, self.dim,
+                          self._sv, self._quasi)
+        self.num_generated += n
+        return self
+
+
+def multinomial_qmc(n_samples, pvals, engine=None, seed=None):
+    """Draw low-discreancy quasi-random samples from multinomial distribution.
+
+    Parameters
+    ----------
+    n_samples : int
+        Number of experiments.
+    pvals: Iterable[float]
+        float vector of probabilities of size ``p``. Elements must be
+        non-negative and sum to 1.
+    engine: QMCEngine
+        Quasi-Monte Carlo engine sampler. If None, Sobol' is used.
+    seed : {int or `numpy.random.RandomState` instance}, optional
+        If `seed` is not specified the `numpy.random.RandomState`
+        singleton is used.
+        If `seed` is an int, a new ``RandomState`` instance is used,
+        seeded with `seed`.
+        If `seed` is already a ``RandomState`` instance, then that
+        instance is used.
+
+    Returns
+    -------
+    samples: array_like (pvals,)
+        int vector of size ``p`` summing to ``n_samples``.
+
+    """
+    if np.min(pvals) < 0:
+        raise ValueError('Elements of pvals must be non-negative.')
+    if not np.isclose(np.sum(pvals), 1):
+        raise ValueError('Elements of pvals must sum to 1.')
+
+    if engine is None:
+        engine = Sobol(1, scramble=True, seed=seed)
+    draws = engine.random(n_samples).ravel()
+    p_cumulative = np.empty_like(pvals, dtype=float)
+    _fill_p_cumulative(np.array(pvals, dtype=float), p_cumulative)
+    sample = np.zeros_like(pvals, dtype=int)
+    _categorize(draws, p_cumulative, sample)
+    return sample
+
+
+class NormalQMC(QMCEngine):
+    """Engine for QMC sampling from a multivariate normal :math:`N(0, I_d)`.
```

Comment:
Yeah `NormalQMC` is a special case of `MultivariateNormalQMC` (note that we get the general case by correlating the samples from this special case with a root decomposition of the covariance matrix. But since we're passing the full covariance in it would be kind of a waste to compute matrix factorizations (as we do in `MultivariateNormalQMC`) on the identity, so special casing this kind of makes sense I feel. But we could do this internally, e.g. by checking whether the matrix is diagonal before computing a factorization.